### PR TITLE
feat: Support OpenSearch plugins (fix #2050)

### LINKED ISSUE #2050: Open search plugins cannot be installed
### Describe the problem and steps to reproduce it:

Relates to this old addons-server issue: https://github.com/mozilla/addons-server/issues/3581

There's no functionality to enable installation of a opensearch add-on in the new mobile pages. 

I'm not sure quite how searches are showing up in the new mobile pages I would have thought the search should just be returning themes and add-ons but they are there if you look for them.

### What happened?

Find an opensearch add-on and try and install it. E.g:  https://addons.mozilla.org/en-US/android/addon/google-dictionary-53151/?src=search

If the fallback button is shown it links to XML as per https://github.com/mozilla/addons-server/issues/3581 -if the addon switch install button is shown it will look like it's installing and then fail.

### What did you expect to happen?

It should install as normal.

The installation needs to be special-cased to work for opensearch in a similar way to the way it works for lightweight themes as mozAddonManager doesn't handle search installs.

### PR DESCRIPTION
**Pull request by tofumatt**
Fixes #2050.

This allows users to install OpenSearch plugins if their browser is compatible (eg is Firefox). Because OpenSearch installs use their own JS API (`window.external.AddSearchProvider`) I opted not to use the `InstallSwitch` component and always use the button for search because we can't really tell when they're installed and the switch + prompt makes for a weird UX in my opinion.

mar-25-2017 14-22-39

### CODE REVIEW

File: tests/client/amo/components/TestAddonDetail.js [hunk 1]
```diff
@@ -313,6 +317,34 @@ describe('AddonDetail', () => {
       rootNode.querySelector('.InstallButton-switch input').disabled);
   });
 
+  it('disables install switch and uses button for OpenSearch plugins', () => {
+    const AddSearchProviderSpy = sinon.stub(window.external,
```
Comment (kumar303) [inline tests/client/amo/components/TestAddonDetail.js]: `window.external` is a non-standard property so I'd rather see a fake `_window` property passed into the component.

File: tests/client/amo/components/TestAddonDetail.js [hunk 2]
```diff
@@ -313,6 +317,34 @@ describe('AddonDetail', () => {
...
+      'AddSearchProvider');
+    const rootNode = renderAsDOMNode({
+      addon: { ...fakeAddon, type: ADDON_TYPE_OPENSEARCH },
+    });
+    const installButton = rootNode.querySelector('.InstallButton-button');
+    assert.equal(installButton.textContent, 'Add to Firefox');
```
Comment (kumar303) [inline tests/client/amo/components/TestAddonDetail.js]: This logic lives in `<InstallButton>` not `<AddonDetail>` so I think it would make more sense to move these tests to `TestInstallButton`

File: tests/client/amo/components/TestAddonDetail.js [hunk 3]
```diff
@@ -313,6 +317,34 @@ describe('AddonDetail', () => {
...
+
+    Simulate.click(installButton);
+
+    assert.ok(AddSearchProviderSpy.calledWith(fakeAddon.installURL));
+  });
+
+  it('disables the button/does nothing if OpenSearch is not available', () => {
```
Comment (kumar303) [inline tests/client/amo/components/TestAddonDetail.js]: Same thing here: I think this should move to `TestInstallButton`

File: tests/client/amo/components/TestAddonDetail.js (hunk 2 above)
Comment (tofumatt) [inline tests/client/amo/components/TestAddonDetail.js]: Oh, oops, you're totally right. Dunno why I put it here. Will move it!

File: tests/client/amo/components/TestAddonDetail.js (hunk 1 above)
Comment (kumar303) [inline tests/client/amo/components/TestAddonDetail.js]: It probably relies on a getter but if you pass in the `window` object like `_window` (or something) then `_window.external` should trigger the getter.

**Comment by tofumatt** [comment]
Sorry, this fell off my radar and I didn't get it fixed. I think I've fixed the issues, ready for another r?

File: src/core/components/InstallButton/index.js [hunk 4]
```diff
@@ -72,6 +84,32 @@ export class InstallButtonBase extends React.Component {
...
+      const onClick = (event) => {
+        event.preventDefault();
+        event.stopPropagation();
+
+        if (
+          !buttonIsDisabled && _window.external &&
+          'AddSearchProvider' in _window.external
+        ) {
+          _log.info('Adding OpenSearch Provider', { addon });
+          _window.external.AddSearchProvider(addon.installURL);
```
Comment (kumar303) [inline src/core/components/InstallButton/index.js]: is there any way to pass a callback or something in case of an error? At the very least we should log errors. But I don't see any way to do it if these docs are accurate: https://developer.mozilla.org/en-US/docs/Web/API/Window/sidebar/Adding_search_engines_from_Web_pages

File: tests/client/core/components/TestInstallButton.js [hunk 5]
```diff
@@ -155,4 +156,72 @@ describe('<InstallButton />', () => {
...
+      _window: fakeWindow,
+    });
+    const installButton = rootNode.querySelector('.InstallButton-button');
+    assert.equal(installButton.textContent, 'Add to Firefox');
+
+    Simulate.click(installButton);
+
+    assert.equal(fakeLog.info.firstCall.args[0], 'Adding OpenSearch Provider');
+    assert.ok(
+      fakeWindow.external.AddSearchProvider.calledWith(fakeAddon.installURL));
```
Comment (kumar303) [inline tests/client/core/components/TestInstallButton.js]: I'd prefer `assert.equal(fakeWindow.external.AddSearchProvider.firstCall.args[0], ...)` because the failure message will be more informative.

File: tests/client/core/components/TestInstallButton.js [hunk 6]
```diff
@@ -155,4 +156,72 @@ describe('<InstallButton />', () => {
...
+  });
+
+  it('disables the button/does nothing if OpenSearch is not available', () => {
+    const fakeLog = { info: sinon.stub() };
+    const fakeWindow = { external: undefined };
```
Comment (kumar303) [inline tests/client/core/components/TestInstallButton.js]: It would be good to add another test for an empty `window.external` like `const fakeWindow = { external: {} };`

File: tests/client/core/test_utils.js [hunk 7]
```diff
@@ -916,6 +917,21 @@ describe('getCompatibleVersions', () => {
...
+      ...fakeAddon,
+      current_version: {
+        compatibility: {},
+      },
+      type: ADDON_TYPE_OPENSEARCH,
+    };
+    getCompatibleVersions({
+      _log: fakeLog, addon: openSearchAddon, clientApp: 'firefox' });
+    assert.include(fakeLog.info.firstCall.args[0],
+      `addon is type ${ADDON_TYPE_OPENSEARCH}`);
```
Comment (kumar303) [inline tests/client/core/test_utils.js]: I don't think checking the log message is so important. I'd rather see this test assert the output of `getCompatibleVersions`. But you could do both.

File: src/core/components/InstallButton/index.js [hunk 8]
```diff
@@ -72,6 +84,32 @@ export class InstallButtonBase extends React.Component {
           {i18n.gettext('Install Theme')}
         </Button>
       );
+    } else if (addon.type === ADDON_TYPE_OPENSEARCH) {
```
Comment (kumar303) [inline src/core/components/InstallButton/index.js]: If `_window.external.AddSearchProvider` is not defined, the button will do nothing which isn't very helpful. Could we disable the button instead or render an error message?

File: src/core/components/InstallButton/index.js (hunk 4 above)
Comment (tofumatt) [inline src/core/components/InstallButton/index.js]: Sadly there's no callback or error event. But if the function doesn't exist it should never be called and it's an API that to my mind only Firefox implements. So it shouldn't ever fail... 🤷‍♀️

File: src/core/components/InstallButton/index.js (hunk 8 above)
Comment (tofumatt) [inline src/core/components/InstallButton/index.js]: Shoot, I thought I disabled it but maybe that got lost when I moved more of the logic into the button? Yes, it should be disabled, I'll do that.

**Comment by tofumatt** [comment]
Okay I've actually reworked this a bit and moved the check into the compatibility util checker function (where it really should have been anyway).

All set for another r?

File: src/amo/components/RatingManager/index.js [hunk 9]
```diff
@@ -97,7 +97,7 @@ export class RatingManagerBase extends React.Component {
         return i18n.gettext('Log in to rate this dictionary');
       case ADDON_TYPE_LANG:
         return i18n.gettext('Log in to rate this language pack');
-      case ADDON_TYPE_SEARCH:
+      case ADDON_TYPE_OPENSEARCH:
         return i18n.gettext('Log in to rate this search engine');
```
Comment (kumar303) [inline src/amo/components/RatingManager/index.js]: Should we call this a 'search plugin?' That's what MDN calls them but sometimes it says 'search engine plugin.'

File: src/core/components/InstallButton/index.js [hunk 10]
```diff
@@ -72,6 +84,29 @@ export class InstallButtonBase extends React.Component {
...
+      const onClick = (event) => {
+        event.preventDefault();
+        event.stopPropagation();
+
+        if (compatible) {
```
Comment (kumar303) [inline src/core/components/InstallButton/index.js]: It looks like this if statement is not necessary since the button will be disabled. If so, I think you should remove the if statement because at least we'd see an error in the console if `onClick` was triggered by a programming mistake -- this would help us track down the bug.

File: src/core/utils.js [hunk 11]
```diff
@@ -385,6 +391,13 @@ export function isCompatibleWithUserAgent({
       return { compatible: false, reason: INCOMPATIBLE_UNDER_MIN_VERSION };
     }
 
+    if (
+      addon.type === ADDON_TYPE_OPENSEARCH &&
+      !(_window.external && 'AddSearchProvider' in _window.external)
+    ) {
+      return { compatible: false, reason: INCOMPATIBLE_NO_OPENSEARCH };
```
Comment (kumar303) [inline src/core/utils.js]: Yeah, this makes way more sense 🌴 🏖

File: tests/client/core/components/TestInstallButton.js [hunk 12]
```diff
@@ -155,4 +161,53 @@ describe('<InstallButton />', () => {
...
+  it('disables the OpenSearch button if not compatible', () => {
+    const root = render({
+      addon: { ...fakeAddon, type: ADDON_TYPE_OPENSEARCH },
+      getClientCompatibility: getClientCompatibilityFalseOpenSearch,
+    });
+
+    assert.equal(root.type, 'div');
+    const buttonComponent = root.props.children[1];
+    assert.equal(buttonComponent.type, Button);
+    assert.equal(buttonComponent.props.className,
```
Comment (kumar303) [inline tests/client/core/components/TestInstallButton.js]: could this just be `assert.include(buttonComponent.props.className, 'InstallButton-button--disabled')` ?

File: src/amo/components/RatingManager/index.js (hunk 9 above)
Comment (tofumatt) [inline src/amo/components/RatingManager/index.js]: Yes actually! "search plugin" I think works (because it's not always searching via a search engine).

File: tests/client/core/components/TestInstallButton.js (hunk 12 above)
Comment (tofumatt) [inline tests/client/core/components/TestInstallButton.js]: Indeed, good call!
